types/mvdan-sh/mvdan-sh-tests.ts

```diff
@@ -1,4 +1,5 @@
-import sh, { LangVariant, Node } from 'mvdan-sh';
+import sh = require('mvdan-sh');
+import { LangVariant, Node } from 'mvdan-sh';
```

Comment:
This will break `import sh from 'mvdan-sh'`?

Maybe we should rewrite the whole definition into commonjs `export = sh`.

Reply:
I'm sorry to say that this won't work.
You cannot export a variable and, alongside it, export named types.

`import sh from 'mvdan-sh'` isn't supposed to work because `mvdan-sh` doesn't have a default export.

`import sh from 'mvdan-sh'` will only (and is only supposed to) work in ESModules.

Reply:
To add a little context here:

In `CommonJS`, TypeScript will transpile a `import sh from "mvdan-sh"` to something comparable to this:
```js
const mvdanSh = require("mvdan-sh");
const sh = mvdanSh.default;
```

This won't work, because `mvdan-sh` doesn't export a `default` member.

A real life example is this here:
<https://www.typescriptlang.org/play?esModuleInterop=false&module=1#code/JYWwDg9gTgLgBAZwBZwGZQiOAiEA3AEwEMA7AWmWwG4AoPIqOMBhAU0YF5EkA6BATxIwiADx4A5VgHcACi3YAKAJRUgA>

Input:
```ts
import sh from "mvdan-sh";
var parser = sh.syntax.NewParser();
```

Output:
```js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mvdan_sh_1 = require("mvdan-sh");
var parser = mvdan_sh_1.default.syntax.NewParser();
```

Take note, that this code breaks and throws a `TypeError`, because `mvdan-sh` doesn't have a default export.

Reply:
It should work if the user uses `esModuleInterop` or `allowSyntheticDefaultImports`.

We can change this definition into something like:

```ts
declare namespace sh {
  export interface ShellScript {}
}

declare const sh: Sh.ShellScript

export = sh
```

I'm on a phone, you can try it by yourself.

Reply:
Declaration merging doesn't work with variables.
Furthermore, the use and the requirement of `esModuleInterop` is not allowed according to this project's readme.

This is most likely the case because every package which relies on a type requiring `esModuleInterop` (even in nested dependencies) would be forced to use `esModuleInterop` as well.

Reply:
It's not allowed to be used in the package definition, not for end user.

`export = sh` is the correct way for commonjs packages.

Reply:
I gotta correct my statement:
According to TypeScript's documentation, declaration merging with namespaces and variables is supposed to work.

However, in this case, it doesn't

It breaks, as soon as I add an enum (in this case the `LangVariant` enum) to the namespace:
https://www.typescriptlang.org/play?esModuleInterop=false&module=1#code/CYUwxgNghgTiAEA7KBbEBnADlMD0At4BvAKHnPhAA9MB7GAF0sQFcV4AZKRAcwDVYAS25NSFcZ248AQlALwAvPAAMAGjITyXXgAUA8gGUAkgA1F8AIzrNWqQFlBMaQYAiAaXqJzAJms3tMlAM6OYAzH7kAL4k0SSgkLAIYLSI6EwEAFzw3ACeANwkJNR0jOYEeUA

Sadly, I don't see another solution as the one I'm currently using in this PR.

Reply:
Awesome! Thank you so much for your time and effort.

Reply:
@manuth

Here we go:

```ts
declare const enum LangVariant {
  LangBash = 0,
  LangPOSIX = 1,
  LangMirBSDKorn = 2,
  LangBats = 3,
}

declare namespace sh {
  interface ShellScript {
    LangVariant: typeof LangVariant
  }
}

declare const sh: sh.ShellScript

export = sh;
```